Addition of the `webextensions.api.tabGroups` documentation (#39370)
mdn/content: files/en-us/mozilla/add-ons/webextensions/api/tabgroups/index.md

---
title: tabGroups
slug: Mozilla/Add-ons/WebExtensions/API/tabGroups
page-type: webextension-api
browser-compat: webextensions.api.tabGroups
---

{{AddonSidebar}}

This API enables extensions to modify and rearrange tab groups.

> 

Reviewer: ```suggestion

The `tabGroups` API does not offer the ability to create or remove tab groups. Use the {{WebExtAPIRef("tabs.group()")}} and {{WebExtAPIRef("tabs.ungroup()")}} methods instead. To query the position of a tab group within a window, use {{WebExtAPIRef("tabs.query()")}}. These APIs in the `tabs` namespace do not require any permissions.
```

## Permissions

To use this API, an extension must request the `"tabGroups"` [optional permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions) in its [`manifest.json`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json) file. The `"tabGroups"` permission is not shown to users in permission prompts.

## Types

- {{WebExtAPIRef("tabGroups.Color")}}
  - : The color of a tab group.
- {{WebExtAPIRef("tabGroups.TabGroup")}}
  - : The state of a tab group.

## Properties

- {{WebExtAPIRef("tabGroups.TAB_GROUP_ID_NONE")}}
  - : The tab group ID value returned when a tab isn't in a tab group.

## Functions

- {{WebExtAPIRef("tabGroups.get()")}}
  - : Returns details about a tab group.
- {{WebExtAPIRef("tabGroups.move()")}}
  - : Moves a tab group within or to another window.
- {{WebExtAPIRef("tabGroups.query()")}}

>   - : Returns all grups or finds tab groups with certain properties.

Reviewer: ```suggestion
  - : Returns all tab groups or finds tab groups with certain properties.
```

- {{WebExtAPIRef("tabGroups.update()")}}
  - : Modifies the state of a tab group.

## Events

- {{WebExtAPIRef("tabGroups.onCreated")}}
  - : Fires when a tab group is created.
- {{WebExtAPIRef("tabGroups.onMoved")}}
  - : Fires when a tab group is moved, within a window or to another window.
- {{WebExtAPIRef("tabGroups.onRemoved")}}
  - : Fires when a tab group is removed.
- {{WebExtAPIRef("tabGroups.onUpdated")}}
  - : Fires when a tab group is updated.

{{WebExtExamples("h2")}}

## Browser compatibility

> {{Compat}}

Reviewer: ```suggestion
{{Compat}}

## See also
- {{WebExtAPIRef("tabs.group()")}}
- {{WebExtAPIRef("tabs.ungroup()")}}
- {{WebExtAPIRef("tabs.query()")}}
- {{WebExtAPIRef("tabs.Tab")}}
```

---
title: tabGroups
slug: Mozilla/Add-ons/WebExtensions/API/tabGroups
page-type: webextension-api
browser-compat: webextensions.api.tabGroups
---

{{AddonSidebar}}

This API enables extensions to modify and rearrange [tab groups](https://support.mozilla.org/en-US/kb/tab-groups).

> Tab groups persist across browser restarts, except for tab groups in private browsing windows.

Reviewer: ```suggestion
Tab groups can persist across browser restarts, except for tab groups in private browsing windows. When a tab group is restored, its tab group ID may differ from its original value.
```

Added additional tweak, because I saw some extension authors tempted to make incorrect assumptions on the stability of the `groupId`.

Author: @Rob--W removing "can" on the basis that, except for those in a private browser window, a tab group is definitely recreated on a browser restart. If the addition of "can" is necessary, we should define the circumstances under which a tab group in a normal window isn't reopened after a browser restart.
```suggestion
Tab groups persist across browser restarts, except for tab groups in private browsing windows. When a tab group is restored, its tab group ID may differ from its original value.
```

> Tab groups persist across browser restarts, except for tab groups in private browsing windows. When a tab group is restored, its tab group ID may differ from its original value.

Reviewer: Continuing thread from https://github.com/mdn/content/pull/39370#discussion_r2083614036

> @Rob--W removing "can" on the basis that, except for those in a private browser window, a tab group is definitely recreated on a browser restart. If the addition of "can" is necessary, we should define the circumstances under which a tab group in a normal window isn't reopened after a browser restart.

Tab groups can persist when the browser session is restored - as the life time of tab groups is tied to the tab that it groups, if tabs are restored on session restore, so are the groups. If the user starts with a new session, then the tab groups are not restored. There is also the concept of a "saved tab group" in both Chrome and Firefox, but these do not have an extension API in either, so I haven't mentioned anything about these.

```suggestion
Tab groups may persist across browser restarts as part of session restore. Tab groups in private browsing windows do not persist across restarts. When a tab group is restored, its `groupId` may differ from its original value.
```

I also included Simeon's suggestion on replacing "tab group ID" with `groupId`.

Author: @Rob--W can we eliminate "may". Ideally, we want a statement of what does or does not happen, not one that implies the behavior is uncertain.
